File: jianzhioffer/firstround/jianzhi_31.py
class Solution(object):
    def validateStackSequences(self, pushed, popped):
        """
        :type pushed: List[int]
        :type popped: List[int]
        :rtype: bool
        """
        if not pushed and not popped:
            return True
        elif not pushed or not popped:
            return False
        query = [pushed.pop(0)]
        while popped:
            if popped[0] not in query:
                query.append(pushed.pop(0))
            elif query[-1] != popped[0]:
                return False
            else:
                query.pop(-1)
                popped.pop(0)
        return True


if __name__ == '__main__':
    p = [1,2,3,4,5]
    q = [[4,3,5,1,2]]
    obj = Solution()
    result = obj.validateStackSequences(p, q)
    print(result)

File: jianzhioffer/secondround/jianzhi_31.cpp
#include <vector>
#include <stack>
using namespace std;

//时间复杂度O(N)，空间复杂度O(N)
class Solution {
public:
    bool validateStackSequences(vector<int>& pushed, vector<int>& popped) {
        if(pushed.empty() && popped.empty()) return true;
        if(pushed.empty() || popped.empty()) return false;
        int push_count = 0;
        int pop_count = 0;
        stack<int> push_stack;
        push_stack.push(pushed[0]);
        push_count++;
        while(pop_count<popped.size())
        {
            if(!push_stack.empty() && push_stack.top()==popped[pop_count])
            {
                push_stack.pop();
                pop_count++;
            }
            else if(push_count<pushed.size())
            {
                push_stack.push(pushed[push_count]);
                push_count++;
            }
            else
                return false;
        }

        return true;
    }
};

int main(int argc, char const *argv[])
{
    vector<int> pushed, popped;
    pushed = {1,0};
    popped = {1,0};
    Solution obj;
    bool result = obj.validateStackSequences(pushed, popped);
    return 0;
}
